Annotate SNV table with mutation frequencies

logstar opened the pull request:
### Purpose/implementation Section

#### What scientific question is your analysis addressing?

Annotate each non-synonymous variant in `snv-consensus-plus-hotspots.maf.tsv.gz` with mutation frequencies per `(cohort, cancer_group, primary/relapse)`.

#### What was your approach?

Subset `snv-consensus-plus-hotspots.maf.tsv.gz` to keep only samples with `sample_type == 'Tumor'`.

Subset `snv-consensus-plus-hotspots.maf.tsv.gz` to keep only non-synonymous variants with the following code.

```R
Variant_Classification %in% c('Frame_Shift_Del',
                              'Frame_Shift_Ins',
                              'Splice_Site',
                              'Nonsense_Mutation',
                              'Nonstop_Mutation',
                              'In_Frame_Del',
                              'In_Frame_Ins',
                              'Missense_Mutation',
                              'Fusion',
                              'Multi_Hit',
                              'Multi_Hit_Fusion',
                              'Hom_Deletion',
                              'Hem_Deletion',
                              'Amp',
                              'Del',
                              'Translation_Start_Site')
```

Create a `Variant_ID` for each variant by concatenating `Chromosome`, `Start_Position`, `Reference_Allele`, and `Tumor_Seq_Allele2` with `'_'`.

Add `Gene_full_name` and `Protein_RefSeq_ID` columns to each variant with annotations obtained from [mygene.info](http://mygene.info/about).

For each `cancer_group`, get each cohort and all cohorts. Call each `cancer_group` and `cohort`(s) combination as a `cancer_group_cohort`. For example,

| cancer_group  | cohort           | n_samples |
|---------------|------------------|-----------|
| Neuroblastoma | CBTN             | 2         |
| Neuroblastoma | GMKF             | 541       |
| Neuroblastoma | TARGET           | 889       |
| Neuroblastoma | CBTN&GMKF&TARGET | 1432      |

For each `cancer_group_cohort` with `n_samples` >= 5, compute `Frequency_in_overall_dataset`, `Frequency_in_primary_tumors`, and `Frequency_in_relapse_tumors` as following:

- `Frequency_in_overall_dataset`:
  - For each unique variant, count the number of patients (identified by `Kids_First_Participant_ID`) that have the variant, and call this number `Total_mutations`.
  - Count the total number of patients in the `cancer_group_cohort`, and call this number `Patients_in_dataset`.
  - `Frequency_in_overall_dataset = Total_mutations / Patients_in_dataset`.

- `Frequency_in_primary_tumors`:
  - For each unique variant, count the number of samples (identified by `Kids_First_Biospecimen_ID`) that are in the `../independent-samples/results/independent-specimens.wgs.primary.tsv`, and call this number `Total_primary_tumors_mutated`.
  - Count the total number of samples in the `cancer_group_cohort` that are also in the `../independent-samples/results/independent-specimens.wgs.primary.tsv`, and call this number `Primary_tumors_in_dataset`.
  - `Frequency_in_primary_tumors = Total_primary_tumors_mutated / Primary_tumors_in_dataset`.

- `Frequency_in_relapse_tumors`:
  - For each unique variant, count the number of samples (identified by `Kids_First_Biospecimen_ID`) that are in the `../independent-samples/results/independent-specimens.wgs.relapse.tsv`, and call this number `Total_relapse_tumors_mutated`.
  - Count the total number of samples in the `cancer_group_cohort` that are also in the `../independent-samples/results/independent-specimens.wgs.relapse.tsv`, and call this number `Relapse_tumors_in_dataset`.
  - `Frequency_in_relapse_tumors = Total_relapse_tumors_mutated / Relapse_tumors_in_dataset`.

Format the SNV mutation frequency table according to the latest spreadsheet that is attached in <https://github.com/PediatricOpenTargets/ticket-tracker/issues/64>.

Merge the SNV mutation frequency tables of all `cancer_group_cohort`s.

#### What GitHub issue does your pull request address?

- <https://github.com/PediatricOpenTargets/ticket-tracker/issues/64>
- <https://github.com/PediatricOpenTargets/ticket-tracker/issues/8>. This issue is no longer compatible with the purpose of this module. This module intends to compute mutation frequencies for each variant, but this issue intents to compute the mutation frequencies for each gene. This issue is listed here for future reference.

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?

The [`get_cg_ch_mut_freq_tbl` function for computing frequencies](https://github.com/logstar/OpenPedCan-analysis/blob/58c7ed0c463f28c0fbf627ebd946b6916ccd4bcc/analyses/snv-frequencies/01-snv-frequencies.R#L81-L252).

The [`get_cg_cs_tbl` function for getting `cacner_group_cohort`s](https://github.com/logstar/OpenPedCan-analysis/blob/58c7ed0c463f28c0fbf627ebd946b6916ccd4bcc/analyses/snv-frequencies/01-snv-frequencies.R#L44-L77).

#### Is there anything that you want to discuss further?

Is it OK to retain the `chr` prefix in the `Chromosome` when creating `Identifer hg38 (See ClinVar ex)` (`Variant_ID` in the result table)? The example value is `10_102599545_G_A` in the spreadsheet, but it will be `chr10_102599545_G_A` in the output table.

I will add RMTL and COSMIC annotations when `CosmicMutantExportCensus.tsv​` and `RMTL_Expanded_3_RMTL_for_OT.csv` are available in future data releases.

I will work on <https://github.com/PediatricOpenTargets/ticket-tracker/issues/82> to add oncoKB annotations to the result table.

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes.

### Results

#### What types of results are included (e.g., table, figure)?

Table

#### What is your summary of the results?

```text
 wc -l results/snv-consensus-annotated-mut-freq.tsv 
  130644 results/snv-consensus-annotated-mut-freq.tsv
```

#### Reproducibility Checklist

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile. **Note: wait to be merged at <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/36>.**
- [ ] This analysis has been added to continuous integration.

#### Documentation Checklist

<!-- Please review https://github.com/AlexsLemonade/OpenPBTA-analysis#documenting-your-analysis -->

- [x] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

<!-- IF YOUR PULL REQUEST IS A DATA RELEASE, PLEASE REMOVE THE [HTML COMMENT TAG](https://html.com/tags/comment-tag/) FROM THE SECTION BELOW AND COMPLETE THE CHECKLIST-->

<!-- 
### Data Release Checklist
- [ ] Is the table in doc/data-file-descriptions.md up to date?
- [ ] Is doc/data-format.md up to date?
- [ ] Is doc/release-notes.md up to date?
- [ ] Is download-data.sh up to date?
- [ ] Was download-data.sh tested and did it complete without error?
-->

logstar commented:
Update to this PR:

**Replace `'NaN%'` frequencies with `''`**

When there is no primary or relapse sample in a `cancer_group_cohort`, mutation frequencies are computed with `0 / 0`, which takes the value of `NaN` in R.

In the result table, replace `'NaN%'` with empty string `''` for the `NaN` frequencies, in order to be consistent with missing values in other columns.

jharenza reviewed (changes requested):
Hi @logstar! Thanks for the quick PR for this table!

Now that #34 is in, will you update this with the files in v6? This will change your CBTN dataset/cohort to PBTA and will condense a few cancer groups. In addition, per our call on Wednesday, we want to add the EFO and MONDO columns to this file, and those can be added with a merge of `efo-mondo-map.tsv`. Note, this is not yet complete - we only have about half of the cancer groups in there so far, but it is a start.

For `CBTN&GMKF&TARGET` in Dataset, can you recode this to `PedOT` - we will use this general designation to mean inclusive of all datasets.

I also noticed that you have some NAs in SIFT, PolyPhen, and some other columns when I read this into R using `read_tsv()`, but it looks like if I do:
```
freq <- read_tsv("~/Documents/GitHub/ped-ot/OpenPedCan-analysis/analyses/snv-frequencies/results/snv-consensus-annotated-mut-freq.tsv",
                 na=character(0))
```
the blanks are retained, so I wanted to note that here because when we do the JSON conversions, which may be in staggered PRs, we will have to add that argument.

Overall, the format looks good, however, I am worried about the results - for instance in neuroblastoma, the expected number of ALK mutations in primary tumors should be ~10%, but we are only seeing ~5.85% and this doesn't align with the current [GMKF pedcbio study](https://pedcbioportal.kidsfirstdrc.org/results/oncoprint?cancer_study_list=x01_fy16_nbl_maris&tab_index=tab_visualize&case_set_id=x01_fy16_nbl_maris_all&Action=Submit&gene_list=ALK&clinicallist=PROFILED_IN_x01_fy16_nbl_maris_cna%2CPROFILED_IN_x01_fy16_nbl_maris_mutations%2CSAMPLE_TYPE) which has a frequency of 9%. There are some mutations in pedcbio we are not capturing here, but I have a suspicion that the study loaded in pedcbio is only using Strelka2 data and not a consensus of 4 callers, so, I have to check into this a bit. Regardless, even with consensus, this number should be higher. cc @migbro @yuankunzhu

logstar commented:
Thank you for the review @jharenza !

> Hi @logstar! Thanks for the quick PR for this table!
> 
> Now that #34 is in, will you update this with the files in v6? This will change your CBTN dataset/cohort to PBTA and will condense a few cancer groups. In addition, per our call on Wednesday, we want to add the EFO and MONDO columns to this file, and those can be added with a merge of `efo-mondo-map.tsv`. Note, this is not yet complete - we only have about half of the cancer groups in there so far, but it is a start.

Yes. I will update this with the v6 data release files and add EFO and MONDO columns.
 
> For `CBTN&GMKF&TARGET` in Dataset, can you recode this to `PedOT` - we will use this general designation to mean inclusive of all datasets.

Sure. Will do.

> I also noticed that you have some NAs in SIFT, PolyPhen, and some other columns when I read this into R using `read_tsv()`, but it looks like if I do:
> 
> ```
> freq <- read_tsv("~/Documents/GitHub/ped-ot/OpenPedCan-analysis/analyses/snv-frequencies/results/snv-consensus-annotated-mut-freq.tsv",
>                  na=character(0))
> ```
>
> the blanks are retained, so I wanted to note that here because when we do the JSON conversions, which may be in staggered PRs, we will have to add that argument.

Good catch! I will add it in the README.md. It will be helpful for anyone who wants to read the tsv table.

For JSON conversion, I was wondering if I could add `jsonlite::write_json(result_tibble, 'result.json')` in the end of the R script to output an additional JSON file, so we can skip the `read_tsv` step and staggered PRs.

> Overall, the format looks good, however, I am worried about the results - for instance in neuroblastoma, the expected number of ALK mutations in primary tumors should be ~10%, but we are only seeing ~5.85% and this doesn't align with the current [GMKF pedcbio study](https://pedcbioportal.kidsfirstdrc.org/results/oncoprint?cancer_study_list=x01_fy16_nbl_maris&tab_index=tab_visualize&case_set_id=x01_fy16_nbl_maris_all&Action=Submit&gene_list=ALK&clinicallist=PROFILED_IN_x01_fy16_nbl_maris_cna%2CPROFILED_IN_x01_fy16_nbl_maris_mutations%2CSAMPLE_TYPE) which has a frequency of 9%. There are some mutations in pedcbio we are not capturing here, but I have a suspicion that the study loaded in pedcbio is only using Strelka2 data and not a consensus of 4 callers, so, I have to check into this a bit. Regardless, even with consensus, this number should be higher. cc @migbro @yuankunzhu

I see a similar ~6.10% ALK overall SNV frequency in my [updated `oncoprint-landscape` module result table](https://github.com/logstar/OpenPedCan-analysis/blob/gene-mut-freq/analyses/oncoprint-landscape/results/primary-plus-Neuroblastoma-mutation_frequency.tsv).

logstar commented:
Updates to this PR:

- **Added EFO, MONDO, and RMTL.** For RMTL, add a single column combining `rmtl` and `version` in the format of `rmtl (version)`, i.e. `Relevant Molecular Target (RMTL version 1)` or `Non-Relevant Molecular Target (RMTL version 1)`.
- **Renamed `cohort` `PBTA&GMKF&TARGET` with `PedOT`.**
- **Updated `README.md` to note `na=character(0)`.**

logstar commented:
Update to this PR:

- **Output SNV mutation frequency table in JSON format**.

Output `m_mut_freq_tbl` in `01-snv-frequencies.R` with the following code:

```R
jsonlite::write_json(
  m_mut_freq_tbl,
  file.path(tables_dir, 'snv-consensus-annotated-mut-freq.json'))
```

`gzip` the JSON file in `run-snv-frequencies.sh` to save space with the following code:

```bash
# --no-name option stops the filename and timestamp from being stored in the
# output file. So rerun will have the same file.
gzip --no-name results/snv-consensus-annotated-mut-freq.json
```

kgaonkar6 commented:
Hi @logstar just wanted to ask if you envision 1 module for each data type like  `snv-frequencies`, `cnv-frequencies` `fusion-frequencies` etc  OR can we rename this module so we can retain frequency/json files here for all data types?

logstar commented:
> Hi @logstar just wanted to ask if you envision 1 module for each data type like `snv-frequencies`, `cnv-frequencies` `fusion-frequencies` etc OR can we rename this module so we can retain frequency/json files here for all data types?

Thank you for asking. I think 1 module for each data type might be more efficient to implement, document, maintain, and review.

If we need to combine the tables into one in the future, we could add another module for combining `snv`, `cnv`, and `fusion` frequency tables together into one.

If there is any reason that we should have one module for all data types, we could also proceed with that.

kgaonkar6 reviewed (commented):
The code overall is well documented  and implements all the checks 👍 

Do you think we could divide the `01` script to have  a function in this module to just gather the counts given a Kids_First_Biospecimen_ID and a Variant_ID/Alt_ID and another `01` script to read mutation/annotation files, call this function and left_join the counts? 

Just wanted to discuss this since I think this would a good way to have reusable function for other data types ( I did play around your code and seem to have one for fusion. which I can add as PR for your review) . Thoughts?

logstar commented:
> The code overall is well documented and implements all the checks 👍
> 
> Do you think we could divide the `01` script to have all your functions in 1 or multiple `utils` scripts in this module to gather the counts given a Kids_First_Biospecimen_ID and a Variant_ID/Alt_ID and another `01` script to read mutation/annotation files, call this function and left_join the counts?
> 
> Just wanted to discuss this since I think this would a good way to have reusable function for other data types ( I did play around your code and seem to have one for fusion. which I can add as PR for your review) . Thoughts?

Thank you for the review @kgaonkar6 !

I agree that we could reuse the workflow design of this module. I will also extract a function to generate the frequency related columns and put it in utils, in order to reuse it to generate the gene-level tables that we discussed this morning. If you need the function immediately, you could adapt the linked code for your own purposes, <https://github.com/logstar/OpenPedCan-analysis/blob/f70645b6c7e4eb15ea29e45e9ebf0adeb5798b9b/analyses/snv-frequencies/01-snv-frequencies.R#L146-L227>.

However, I woud prefer to use the code from other modules by adapting them rather than directly resuing via `source('../another-module/utils/some_function.R')`, so I do not have rely on any code that has no guarantee to not change interface in the future.

kgaonkar6 commented:
That's good point! Since the tables/json generation is still in progress for each data type I agree that adapting the code structure in different modules would be better 👍

logstar commented:
Update to this PR:

Changed independent sample lists to v6 `independent-specimens.wgs.primary.eachcohort.tsv` and `independent-specimens.wgs.relapse.eachcohort.tsv`, which are obtained from the [recently updated independent-samples module](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/48).

Previously used independent sample list is v5 `independent-specimens.wgs.primary.tsv` and `independent-specimens.wgs.relapse.tsv`.

In v6, the differences between all-cohorts and each-cohort independent sample lists are:

- Number of samples:

```text
$ wc -l ../independent-samples/results/independent-specimens.wgs.{primary.,relapse.}*tsv
 1010 ../independent-samples/results/independent-specimens.wgs.primary.eachcohort.tsv
 1006 ../independent-samples/results/independent-specimens.wgs.primary.tsv
  172 ../independent-samples/results/independent-specimens.wgs.relapse.eachcohort.tsv
  168 ../independent-samples/results/independent-specimens.wgs.relapse.tsv
```

<details><summary>- Patient and sample IDs</summary>
<p>

```text
$ diff <(sort ../independent-samples/results/independent-specimens.wgs.primary.tsv) <(sort ../independent-samples/results/independent-specimens.wgs.primary.eachcohort.tsv)
389a390
> PT_2E552BAR   BS_6Z213H2V
450c451
< PT_59D00MBQ   BS_3E5C1PN1
---
> PT_59D00MBQ   BS_QG6V29H7
497a499
> PT_7JQ24F35   BS_895FJ34T
539c541
< PT_9GKVQ9QS   BS_JRFVST47
---
> PT_9GKVQ9QS   BS_3Z40EZHD
671c673
< PT_FZHGKJ0H   BS_NHMPQ8Q1
---
> PT_FZHGKJ0H   BS_5J5VH3X0
696c698
< PT_H45M7M2T   BS_E9M7TDB6
---
> PT_H45M7M2T   BS_6XHDR1WS
745c747
< PT_K8ZV7APT   BS_EXTEGB51
---
> PT_K8ZV7APT   BS_1RFBH1SP
795c797
< PT_NESAQHB1   BS_KJTW1RBN
---
> PT_NESAQHB1   BS_6TMPZKSZ
868a871
> PT_S4YNE17X   BS_3MYQP7P5
869a873
> PT_S4YNE17X   BS_CRKBDAYZ
```

```text
$ diff <(sort ../independent-samples/results/independent-specimens.wgs.relapse.tsv) <(sort ../independent-samples/results/independent-specimens.wgs.relapse.eachcohort.tsv)
8c8,9
< PT_1H2REHT2   BS_79SYEHY3
---
> PT_1H2REHT2   BS_7M7JNG00
> PT_1H2REHT2   BS_96S0VQBN
23c24,25
< PT_2WVW55DA   BS_RG6JK7ZN
---
> PT_2WVW55DA   BS_M28HG0NT
> PT_2WVW55DA   BS_R94EDP5H
25c27
< PT_2YT37G8P   BS_ZJZFT0AJ
---
> PT_2YT37G8P   BS_4QFSH7C4
44c46
< PT_8GN3TQRM   BS_S791VC80
---
> PT_8GN3TQRM   BS_QQP17PDQ
48c50
< PT_962TCBVR   BS_ME97J1GA
---
> PT_962TCBVR   BS_Q3ANJDVG
58c60
< PT_AQWDQW27   BS_D3KVGMRJ
---
> PT_AQWDQW27   BS_JR56RQDC
70c72,73
< PT_DH8A0SNQ   BS_B5Z2XV03
---
> PT_DFQAH7RS   BS_ZKFEMJZ8
> PT_DH8A0SNQ   BS_BYJDRXMK
101a105
> PT_KTRJ8TFY   BS_EE73VE7V
103,104c107,108
< PT_KZ56XHJT   BS_0ATJ22QA
< PT_MDWPRDBT   BS_D368BNRD
---
> PT_KZ56XHJT   BS_YHXMYDBN
> PT_MDWPRDBT   BS_00TRPEQX
137c141
< PT_S2SQJVGK   BS_YGQS4660
---
> PT_S2SQJVGK   BS_8S2CZBA6
139c143
< PT_S4YNE17X   BS_8CA214YR
---
> PT_S4YNE17X   BS_D3XVXF05
167c171
< PT_ZZRBX5JT   BS_JE1DF581
---
> PT_ZZRBX5JT   BS_9N3B3HZB
```

</p>

logstar commented:
Update to this PR:

Converted `results/snv-consensus-annotated-mut-freq.json` to JSONL format with `jq`. For more details, see <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/27#issuecomment-876103781>.

logstar commented:
@jharenza I wonder if we need to change the `Dataset` column values that have more than one cohorts to `all_cohorts`, in order to be consistent with [the TPM tables](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/27#issuecomment-876007589).

jharenza commented:
> @jharenza I wonder if we need to change the `Dataset` column values that have more than one cohorts to `all_cohorts`, in order to be consistent with [the TPM tables](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/27#issuecomment-876007589).
> 
> Currently, the unique `Dataset` values are as following, with PedOT = PBTA&GMKF&TARGET.
> 
> ```
> GMKF
> PBTA
> PedOT
> ```

yes, we should do this for all modules that create tables - i was thinking maybe we need some documentation about minimum fields for the modules so we can keep everything consistent and straight - thoughts?

logstar commented:
> > @jharenza I wonder if we need to change the `Dataset` column values that have more than one cohorts to `all_cohorts`, in order to be consistent with [the TPM tables](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/27#issuecomment-876007589).
> > Currently, the unique `Dataset` values are as following, with PedOT = PBTA&GMKF&TARGET.
> > ```
> > GMKF
> > PBTA
> > PedOT
> > ```
> 
> yes, we should do this for all modules that create tables - i was thinking maybe we need some documentation about minimum fields for the modules so we can keep everything consistent and straight - thoughts?

Thank you for the quick reply. I will change accordingly.

Regarding minimum fields, I think we need at least the following three fields for describing what and how samples are used for computing certain values or generating certain plots.

- Disease/cancer_group
- Dataset/cohort
- The way how all samples are grouped together, which is currently implicitly inferred from Dataset/cohort. In the TPM and SNV modules, this field is only necessary for the TPM mean gene-wise z-scores, because the TPM mean gene-wise z-scores for a cancer_group and cohort may change if other samples are grouped differently.

I was using a `sample_group` concept when developing the TPM tables, which describes cancer_group, cohort, and the grouping method for all samples.

Following is the explanation of the `sample_group` column that I wrote for a PR update <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/27#issuecomment-874812735>.

>#### Create an additional `sample_group` column for grouping samples in all-cohorts and each-cohort analyses
>
>The `sample_group` column needs to describe three things of a sample:
>
>- `cancer_group`
>- `cohort`(s)
>- The way how all samples are grouped together, either by each-`cancer_group`-all-`cohort`s or each-`cancer_group`-each-`cohort`, which are referred as `all-cohorts` and `each-cohort` methods respectively in this analysis. The gene-wise z-scores for the same (`cancer_group`, `cohort`) may be different when computed using the `all-cohorts` or `each-cohort` grouping method, if other samples are not identically grouped together. For example, in the following groups, `all-cohorts-g3` and `each-cohort-g5` may have different gene-wise zscores, even though they both contain the same samples that have (`cancer_group_C`, `cohort_Z`).
>
>  | `cancer_group`   | `cohort`   | `all-cohorts` group IDs | `each-cohort` group IDs |
>  |------------------|------------|-------------------------|-------------------------|
>  | `cancer_group_A` | `cohort_X` | `g1`                    | `g1`                    |
>  | `cancer_group_A` | `cohort_Y` | `g1`                    | `g2`                    |
>  | `cancer_group_B` | `cohort_X` | `g2`                    | `g3`                    |
>  | `cancer_group_B` | `cohort_Y` | `g2`                    | `g4`                    |
>  | `cancer_group_C` | `cohort_Z` | `g3`                    | `g5`                    |
>
>For backward compatibility of the results:
>
>- Specific cohort(s) are implicit in the `all-cohorts` grouping method.
>- Grouping method is implicit in the `each-cohort` grouping method.

logstar commented:
Update to this PR:

Added gene-level mutation frequency table. Previous variant-level tables are renamed with additional `var-level-` prefix.

This is a work in progress for addressing [issue #91](https://github.com/PediatricOpenTargets/ticket-tracker/issues/91). Need to add the additional `gene_type` column.

logstar commented:
Update to this PR:

Changed `PedOT` to `all_cohorts` for `Dataset` column values with more than one cohorts.

logstar commented:
Update to this PR:

Added a single `Gene_type` column to the gene-level SNV frequency table. The `Gene_type` information is obtained from `../fusion_filtering/references/genelistreference.txt`, and its sources are described at <https://github.com/d3b-center/annoFuse#prerequisites-for-cohort-level-analysis>.

logstar commented:
## Updates to this PR

### Added PedcBio oncoprint and mutations plot URLs to the mutational frequency tables

The `case_set_id`s (aka `sampleListId`s and `Case List`s) of [PedcBio Pediatric Open Targets study](https://pedcbioportal.kidsfirstdrc.org/study/summary?id=ped_opentargets_2021) are used to link `cancer_group` and `cohort` in [PediatricOpenTargets/OpenPedCan-analysis](https://github.com/PediatricOpenTargets/OpenPedCan-analysis) data releases. The `case_set_id`s of the PedcBio Pediatric Open Targets study is created by @migbro as described in <https://github.com/PediatricOpenTargets/ticket-tracker/issues/93>.

Example URLs:

- <https://pedcbioportal.kidsfirstdrc.org/results/oncoprint?cancer_study_list=ped_opentargets_2021&case_set_id=ped_opentargets_2021_pbta_adamantinomatous_craniopharyngioma&Action=Submit&gene_list=DNAH5>
- <https://pedcbioportal.kidsfirstdrc.org/results/mutations?cancer_study_list=ped_opentargets_2021&case_set_id=ped_opentargets_2021_pbta_adamantinomatous_craniopharyngioma&Action=Submit&gene_list=DNAH5>
- <https://pedcbioportal.kidsfirstdrc.org/results/oncoprint?cancer_study_list=ped_opentargets_2021&case_set_id=ped_opentargets_2021_pbta_high-grade_glioma_astrocytoma&Action=Submit&gene_list=CLPSL2>
- <https://pedcbioportal.kidsfirstdrc.org/results/mutations?cancer_study_list=ped_opentargets_2021&case_set_id=ped_opentargets_2021_pbta_high-grade_glioma_astrocytoma&Action=Submit&gene_list=CLPSL2>

Issue addressed: <https://github.com/PediatricOpenTargets/ticket-tracker/issues/83>

### Changed `wgs` primary/relapse independent sample lists to `wgswxspanel` ones for computing mutation frequencies

`snv-consensus-plus-hotspots.maf.tsv.gz` has 1240 WGS, 31 WXS, and 1 Targeted Sequencing samples, so `wgswxspanel` primary/relapse independent sample lists should be used to include all samples for computing mutation frequencies.

@kgaonkar6 @ewafula This change, which is implemented in <https://github.com/logstar/OpenPedCan-analysis/commit/5e917eb043be748ebff30125867e91c81e46d3e1>, might also be necessary for [`fusion-frequencies`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/49) and [`cnv-frequencies`](https://github.com/PediatricOpenTargets/ticket-tracker/issues/66) modules.

### Used only overlapping tumor samples between `snv-consensus-plus-hotspots.maf.tsv.gz` and `histologies.tsv` for computing mutation frequencies

Previously, all tumor samples in `snv-consensus-plus-hotspots.maf.tsv.gz` and `histologies.tsv` are used for computing mutation frequencies.

However, the tumor samples in `snv-consensus-plus-hotspots.maf.tsv.gz` and the tumor samples in `histologies.tsv` do not fully overlap.

Therefore, only the overlapping samples between `snv-consensus-plus-hotspots.maf.tsv.gz` and `histologies.tsv` should be used for computing mutation frequencies.

This change is implemented in <https://github.com/logstar/OpenPedCan-analysis/commit/c07b2b569c89dec167ddf6125c33c34ee0fb3b19>.

@kgaonkar6 @ewafula Similar changes might also be necessary for [`fusion-frequencies`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/49) and [`cnv-frequencies`](https://github.com/PediatricOpenTargets/ticket-tracker/issues/66) modules.

kgaonkar6 reviewed (commented):
Hi @logstar thanks for all the updates. The code generates variant and gene level tables and adds gene annotation (RMTL, Gene full name, Ensembl gene and protein  ID,Gene_type) and disease annotation (EFO, Mondo) as well as pedcbio links!
I have reviewed that the code runs perfectly creating the desired outputs ( except the conversion from json to jsonlines, did we send this to the OT portal folks for verification?) 

Any chance the `analyses/snv-frequencies/01-snv-frequencies.R` can be broken into smaller chunks of code? I think 1000 lines of code is a lot to be able to review thoroughly and potentially might intimidate anyone that would be asked to update/maintain the code later. My suggestion  ( feel free to take or leave it ) is put functions in a `utils` folder and divide the variant frequency table generation and gene based frequency table generation into 2 scripts. 

As part OpenPBTA peer review process we would keep lines of code max ~ 400 and 1 script would focus on 1 task.

logstar commented:
> Hi @logstar thanks for all the updates. The code generates variant and gene level tables and adds gene annotation (RMTL, Gene full name, Ensembl gene and protein ID,Gene_type) and disease annotation (EFO, Mondo) as well as pedcbio links!
> I have reviewed that the code runs perfectly creating the desired outputs ( except the conversion from json to jsonlines, did we send this to the OT portal folks for verification?)
> 
> Any chance the `analyses/snv-frequencies/01-snv-frequencies.R` can be broken into smaller chunks of code? I think 1000 lines of code is a lot to be able to review thoroughly and potentially might intimidate anyone that would be asked to update/maintain the code later. My suggestion ( feel free to take or leave it ) is put functions in a `utils` folder and divide the variant frequency table generation and gene based frequency table generation into 2 scripts.
> 
> As part OpenPBTA peer review process we would keep lines of code max ~ 400 and 1 script would focus on 1 task.

Thank you for the review @kgaonkar6 .

The jsonlines files have not yet been verified by the OT portal folks. The files will probably be verified in the next iteration.

I have considered putting functions into a `utils` since the last time you suggested, but I think putting all functions in the same file as the input/output/computing procedure would be more informative for both reviewers and maintainers. As `utils` or any external functions implies that they are usually not affected by the data formats, e.g. column names, PedcBio `case_set_id` composition, and additional annotations, putting all functions in the same file as the input/output/computing procedure would be less surprising for future maintainers to find out that most functions also need to be revised for any update on the results. Similarly, putting all functions in one file also informs the reviewer that all functions need to be evaluated for any updates in the future.

I have also considered creating separate gene-level and var-level scripts, but I decided to put them in the same file to be more efficient, especially for future updates with a lot more WGS, WXS and panel samples. The most time consuming step in this module is reading the MAF file, which is already taking a couple of minutes for v6 data, so creating two files will double the runtime of this module, which significantly delays any trial and error in future development.

I agree that > 1000 lines of code is a lot of work for any update or review, but I think it is a good trade-off for being more informative and efficient. In addition, separating functions into external files will not really reduce the workload for future reviewers or maintainers.

logstar commented:
Updates to this PR:

- **Added OncoKB cancer gene and oncogene/TSG columns to the gene-level and var-level mutation frequency tables.**

Added column names are `OncoKB_cancer_gene` and `OncoKB_oncogene_TSG`.

The OncoKB gene annotation data `input/oncokb_cancer_gene_list.tsv` is downloaded at <https://www.oncokb.org/cancerGenes>, which does not require OncoKB API license.

If we need to add OncoKB annotations that are not included in `input/oncokb_cancer_gene_list.tsv` in the future, the `histologies.tsv` needs an additional column `ONCOTREE_CODE`, which is required by oncokb-annotator at <https://github.com/oncokb/oncokb-annotator/blob/8e0d7aded0b4312ed8fa42e34842f6e1d71797a4/MafAnnotator.py#L36>. More information about `ONCOTREE_CODE` is at <http://oncotree.mskcc.org>.

- **Added a guide how to update `01-snv-frequencies.R`.**

It is concerning that `01-snv-frequencies.R` would be hard to maintain and review, as raised by @kgaonkar6 at <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/45#pullrequestreview-705519113>, so I added a guide in the `README.md` on how to update `01-snv-frequencies.R`.

The guide has the following content:

- Summary of different code sections.
- Descriptions of how to perform common updates.
- Explain how to use `stopifnot()` statements and function test cases.

jharenza commented:
> The OncoKB gene annotation data `input/oncokb_cancer_gene_list.tsv` is downloaded at https://www.oncokb.org/cancerGenes, which does not require OncoKB API license.

Great! Can we just confirm we can distribute this list as well?

logstar commented:
> > The OncoKB gene annotation data `input/oncokb_cancer_gene_list.tsv` is downloaded at https://www.oncokb.org/cancerGenes, which does not require OncoKB API license.
> 
> Great! Can we just confirm we can distribute this list as well?

Sure. I will check with the OncoKB team.

logstar commented:
> > The OncoKB gene annotation data `input/oncokb_cancer_gene_list.tsv` is downloaded at https://www.oncokb.org/cancerGenes, which does not require OncoKB API license.
> 
> Great! Can we just confirm we can distribute this list as well?

I confirmed with the OncoKB group that we can make the `OncoKB_cancer_gene` and `OncoKB_oncogene_TSG` information publicly accessible in the PedOT database. Just to note here, the OncoKB group suggested that we should give proper credits to the OncoKB in the PedOT database.

logstar commented:
Added WIP label to this PR. I will add unit testing framework using the `testthat`package, in order to remove the test cases in `01-snv-frequencies.R`, like the one that is implemented in <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/55>.

logstar commented:
@jharenza @kgaonkar6 I have added unit testing for this PR and removed test case comments in `01-snv-frequencies.R`.

All tests passed in the Docker image. See the commits from <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/45/commits/c58c8d596aa06a424175e50d2764b20c600b2c17> for more details.

```text
$ bash run-tests.sh
✔ |  OK F W S | Context
✔ |  23       | tests/test_get_cohort_set_value.R
✔ |  31       | tests/test_get_opr_mut_freq_tbl.R [0.8 s]
✔ |  33       | tests/test_get_pcb_pot_csi.R
✔ |  42       | tests/test_get_pcb_pot_plot_url.R
✔ |  21       | tests/test_helper_import_function.R
✔ |  12       | tests/test_num_to_pct_chr.R

══ Results ═════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 1.1 s

OK:       162
Failed:   0
Warnings: 0
Skipped:  0
Done running run-tests.sh
```

I will work on <https://github.com/PediatricOpenTargets/ticket-tracker/issues/122> before incorporating the `annotator` API into this PR. If this PR is merged, I will create a new PR for refactoring the annotation part in this module.

logstar commented:
Added WIP label to this PR.

I will use the [`annotator` API](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/long-format-table-utils) to add annotations to the SNV tables. This will reduce the number of lines in `analyses/snv-frequencies/01-snv-frequencies.R`, which may reduce the workload for the reviewers to review this PR.

I will keep using v6 data and independent sample lists in this PR, and I will create a ticket for rerunning the `snv-frequency` module using the v7 data and independent sample lists.

jharenza commented on analyses/README.md:
```diff
@@ -53,6 +53,7 @@ Note that _nearly all_ modules use the harmonized clinical data file (`pbta-hist
 | [`selection-strategy-comparison`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/selection-strategy-comparison) | `pbta-gene-expression-rsem-fpkm.polya.rds` <br> `pbta-gene-expression-rsem-fpkm.stranded.rds` | *Deprecated*; Comparison of RNA-seq data from different selection strategies | N/A
 | [`sex-prediction-from-RNASeq`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/sex-prediction-from-RNASeq) | `pbta-gene-expression-kallisto.stranded.rds` <br> `pbta-histologies.tsv` | *In progress*; predicts genetic sex using RNA-seq data ([#84](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/6)) | N/A
 | [`snv-callers`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/snv-callers) | `pbta-snv-lancet.vep.maf.gz` <br> `pbta-snv-mutect2.vep.maf.gz` <br> `pbta-snv-strelka2.vep.maf.gz` <br> `pbta-snv-vardict.vep.maf.gz` <br> `tcga-snv-lancet.vep.maf.gz` <br> `tcga-snv-mutect2.vep.maf.gz` <br> `tcga-snv-strelka2.vep.maf.gz` | Generates consensus SNV and indel calls for PBTA and ...
+| [`snv-frequencies`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/snv-frequencies) | `histologies.tsv` <br> `snv-consensus-plus-hotspots.maf.tsv.gz` <br> `independent-samples/results/independent-specimens.wgs.primary.tsv` <br> `independent-samples/results/independent-specimens.wgs.relapse.tsv` | Annotate SNV table with mutation frequencies | N/A
```
Just a note that you will want to add the WXS data with v7 as well

logstar replied:
Thank you for the note. This is good to know! I will use `wgswxspanel` independent sample lists in the upcoming snv-frequencies v7 update PR.

jharenza commented on analyses/snv-frequencies/run-snv-frequencies.sh:
```diff
@@ -0,0 +1,63 @@
+# - https://stackoverflow.com/a/48711608/4638182
+# - https://stackoverflow.com/a/66709708/4638182
+jq --compact-output '.[]' \
+  results/var-level-snv-consensus-annotated-mut-freq.json \
+  > results/var-level-snv-consensus-annotated-mut-freq.jsonl
+
+jq --compact-output '.[]' \
+  results/gene-level-snv-consensus-annotated-mut-freq.json \
+  > results/gene-level-snv-consensus-annotated-mut-freq.jsonl
+
+
+# --no-name option stops the filename and timestamp from being stored in the
+# output file. So rerun will have the same file.
+gzip --no-name results/var-level-snv-consensus-annotated-mut-freq.json
+gzip --no-name results/var-level-snv-consensus-annotated-mut-freq.jsonl
+
+gzip --no-name results/gene-level-snv-consensus-annotated-mut-freq.json
+gzip --no-name results/gene-level-snv-consensus-annotated-mut-freq.jsonl
```
I think you can remove the `*.json.gz` files so we don't have to store bulk redundant files and since jsonl is the desired format.

jharenza replied:
Also, minor comment to write out `variant-level` here

logstar replied:
Thank you for the suggestions.

Removed `*.json.gz` files in commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/45/commits/495803998e893b128fbe5070d7ef8fb25b30732b>.

Write out `variant-level` in filenames in commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/45/commits/3b5e6354dfb6ca85a74e855b4a1878c740d56810>.

jharenza commented on analyses/snv-frequencies/01-snv-frequencies.R:
```diff
@@ -0,0 +1,1037 @@
+# - valid_url_case_set_ids: a vector of character values that are valid
+#   case_set_ids for generating PedcBio PedOT oncoprint and mutations plot URLs.
+#
+# Returns a MAF tibble with additional frequency columns.
+get_cg_ch_var_level_mut_freq_tbl <- function(maf_df, overall_histology_df,
+                                             primary_histology_df,
+                                             relapse_histology_df,
+                                             ss_cancer_group, ss_cohorts,
+                                             valid_url_case_set_ids) {
```
For now, we can leave this in, but I am wondering if since we have the gene level table, we should remove the gene level plotting in the variant level table because it may be confusing since it is not the exact set of samples represented in the percentage in the table. We can ask tomorrow.

logstar replied:
I will keep a note of this. If necessary, I will remove the pedcbio plotting URLs from the variant-level tables in the upcoming snv-frequencies v7 update PR.

jharenza reviewed (approved):
Really nice job on this - two minor comments about removing redundant json files and renaming the variant level file, and once addressed, this can be merged.
